client/mm/exchange_adaptor.go

```diff
@@ -2638,6 +2498,412 @@ func (u *unifiedExchangeAdaptor) handleDEXNotification(n core.Notification) {
+		sells := utils.Min(dexBaseLots, cexQuoteLots, maxSellLots)
+		buys := utils.Min(dexQuoteLots, cexBaseLots, maxBuyLots)
+		return buys + sells
+	}
+
+	currentScore := matchability(baseInv.dexLots, quoteInv.dexLots, baseInv.cexLots, quoteInv.cexLots)
+
+	// targetedSplit finds a distribution that targets a specified ratio of
+	// dex-to-cex.
+	targetedSplit := func(avail, dexTarget, cexTarget, dexLot, cexLot uint64) (dexLots, cexLots, extra uint64) {
+		if dexTarget+cexTarget == 0 {
+			return
+		}
+		cexR := float64(cexTarget*cexLot) / float64(dexTarget*dexLot+cexTarget*cexLot)
+		cexLots = uint64(math.Round(cexR*float64(avail))) / cexLot
+		cexBal := cexLots * cexLot
+		dexLots = (avail - cexBal) / dexLot
+		dexBal := dexLots * dexLot
+		if cexLot < dexLot {
+			cexLots = (avail - dexBal) / cexLot
+			cexBal = cexLots * cexLot
+		}
+		extra = avail - cexBal - dexBal
+		return
+	}
+
+	baseSplit := func(dexTarget, cexTarget uint64) (dexLots, cexLots, extra uint64) {
+		return targetedSplit(baseAvail, utils.Min(dexTarget, maxSellLots), utils.Min(cexTarget, maxBuyLots), perLot.dexBase, perLot.cexBase)
+	}
+	quoteSplit := func(dexTarget, cexTarget uint64) (dexLots, cexLots, extra uint64) {
+		return targetedSplit(quoteAvail, utils.Min(dexTarget, maxBuyLots), utils.Min(cexTarget, maxSellLots), perLot.dexQuote, perLot.cexQuote)
+	}
+
+	// We'll keep track of any distributions that have a matchability score
+	// better than the score for the current distribution.
+	type scoredSplit struct {
+		score uint64 // matchability
+		// spread is just the minimum of baseDeposit, baseWithdraw, quoteDeposit
+		// and quoteWithdraw. This is tertiary criteria for prioritizing splits,
+		// with a higher spread being preferable.
+		spread                      uint64
+		fees                        uint64
+		baseDeposit, baseWithdraw   uint64
+		quoteDeposit, quoteWithdraw uint64
+	}
+	baseSplits := [][2]uint64{
+		{baseInv.dex, baseInv.cex},   // current
+		{dexSellLots, dexBuyLots},    // ideal
+		{quoteInv.cex, quoteInv.dex}, // match the counter asset
+	}
+	quoteSplits := [][2]uint64{
+		{quoteInv.dex, quoteInv.cex},
+		{dexBuyLots, dexSellLots},
+		{baseInv.cex, baseInv.dex},
+	}
+
+	splits := make([]*scoredSplit, 0)
+	// scoreSplit gets a score for the proposed asset distribution and, if the
+	// score is higher than currentScore, saves the result to the splits slice.
+	scoreSplit := func(dexBaseLots, dexQuoteLots, cexBaseLots, cexQuoteLots, extraBase, extraQuote uint64) {
+		score := matchability(dexBaseLots, dexQuoteLots, cexBaseLots, cexQuoteLots)
+		if score <= currentScore {
+			return
+		}
+
+		var fees uint64
+		var baseDeposit, baseWithdraw, quoteDeposit, quoteWithdraw uint64
+		if dexBaseLots != baseInv.dexLots || cexBaseLots != baseInv.cexLots {
+			fees++
+			dexTarget := dexBaseLots*perLot.dexBase + additionalBaseFees + extraBase
+			cexTarget := cexBaseLots * perLot.cexBase
```

Comment:
Why does `extraBase` all go to the dex, but `extraQuote` is split between the dex and the cex.

Reply:
Slippage

Reply:
This comment made me realize though that we could send a little less base asset to the cex. Damn.

Reply:
Or maybe you're asking because you might want to keep extra on the cex for later once you shut the bot down?

Reply:
I think the CEX could also use a buffer, in case the arbs go in the same direction multiple times, to avoid having to rebalance again.

Reply:
When we buy on dex, our corresponding sell on cex is a perfect multiple of the lot size. Only the quantity of quote asset varies with the rate. The base asset is always in lots. I think it's just a consequence of the dex protocol defining lot sizes in terms of the base asset. I could be confused but I'll try to sketch it out here.
buy -> sell: buy n * lotSize dcr at rate r, sell n * lotSize dcr at rate (r + profit + feeGap), ending up with more btc than we started with
sell -> buy: sell n * lotSize dcr at rate r to receive r * n * lotSize btc, sell (r - profit - feeGap) * n * lotSize btc on cex, netting btc again.

Reply:
That's true, but let's say, for example, our configuration has 5 sell lots and 5 buy lots, and we have 14 lots of BTC. If we put 9 on the DEX and 5 on the CEX, then we sell one on the CEX, we now have to rebalance again. It's better to split it evenly to hopefully avoid future rebalances.

Reply:
>  If we put 9 on the DEX and 5 on the CEX...

That shouldn't happen, since we will be targeting a 50 / 50 split. We're not splitting up the 4 lots above the required 10, we're splitting up the < 1 lot change after we split 7/7. If we start with 14.5 xyz, lot size is 1, the "extra" is the 0.5.

Reply:
OK I understand now, thanks.